timemachine/cpp/src/barostat.cu

```diff
@@ -0,0 +1,328 @@
+    gpuErrchk(cudaFree(d_mol_offsets_));
+    gpuErrchk(cudaFree(d_box_after_));
+    gpuErrchk(cudaFree(d_u_before_));
+    gpuErrchk(cudaFree(d_u_after_));
+};
+
+
+void __global__ rescale_positions(
+    const int N, // Number of atoms to shift
+    double * __restrict__ coords, // Cordinates
+    const double length_scale,
+    const double * __restrict__ box, // [9]
+    double * __restrict__ scaled_box, // [9]
+    const int * __restrict__ atom_idxs, // [N]
+    const int * __restrict__ mol_idxs, // [N]
+    const int * __restrict__ mol_offsets, // [N]
+    const double * __restrict__ centroids // [N*3]
+) {
+    const int idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (idx >= N) {
+        return;
+    }
+    const int atom_idx = atom_idxs[idx];
+    const int mol_idx = mol_idxs[idx];
+
+    double center_x = box[0*3+0] * 0.5;
+    double center_y = box[1*3+1] * 0.5;
+    double center_z = box[2*3+2] * 0.5;
+
+    const double num_atoms = static_cast<double>(mol_offsets[mol_idx+1] - mol_offsets[mol_idx]);
+
+    const double centroid_x = centroids[mol_idx*3+0] / num_atoms;
+    const double centroid_y = centroids[mol_idx*3+1] / num_atoms;
+    const double centroid_z = centroids[mol_idx*3+2] / num_atoms;
+
+    const double displacement_x = ((centroid_x - center_x) * length_scale) + center_x - centroid_x;
+    const double displacement_y = ((centroid_y - center_y) * length_scale) + center_y - centroid_y;
+    const double displacement_z = ((centroid_z - center_z) * length_scale) + center_z - centroid_z;
+
+    coords[atom_idx*3+0] += displacement_x;
+    coords[atom_idx*3+1] += displacement_y;
+    coords[atom_idx*3+2] += displacement_z;
+    if (atom_idx == 0) {
+        scaled_box[0*3+0] *= length_scale;
+        scaled_box[1*3+1] *= length_scale;
+        scaled_box[2*3+2] *= length_scale;
+    }
+}
+
+void __global__ find_group_centroids(
+    const int N, // Number of atoms to shift
+    const double * __restrict__ coords, // Coordinates
+    const int * __restrict__ atom_idxs, // [N]
+    const int * __restrict__ mol_idxs, // [N]
+    double * __restrict__ centroids // [num_molecules * 3]
+) {
+    const int idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (idx >= N) {
+        return;
+    }
+    const int atom_idx = atom_idxs[idx];
+    const int mol_idx = mol_idxs[idx];
+    atomicAdd(centroids + mol_idx*3+0, coords[atom_idx*3+0]);
+    atomicAdd(centroids + mol_idx*3+1, coords[atom_idx*3+2]);
+    atomicAdd(centroids + mol_idx*3+2, coords[atom_idx*3+2]);
+}
+
+void MonteCarloBarostat::reset_counters() {
+    num_attempted_ = 0;
+    num_accepted_ = 0;
+}
+
+void MonteCarloBarostat::inplace_move(
+    double *d_x,
+    double *d_box,
+    const double lambda
+    ) {
+
+    step_++;
+    if (step_ % interval_ != 0) {
+        return;
+    }
+    std::vector<double> h_box(9);
+    gpuErrchk(cudaMemcpy(&h_box[0], d_box, 9*sizeof(d_box), cudaMemcpyDeviceToHost));
+
+    double volume = h_box[0*3+0]*h_box[1*3+1]*h_box[2*3+2];
```

Comment:
Not sure if you folks will only ever support rectilinear boxes, but this volume expression will only work for rectilinear boxes (where all box angles are`pi/2`). You can simply replace this with `abs(det(h_box))` if you want it to be fully general for paralellepipeds.